쥬스메이커 [STEP 1] Gundy, 준호

Gundy93 opened the pull request:
@junbangg 
안녕하세요 알라딘!🏜
건디와 준호의 첫 PR입니다.🥹
여러모로 고민이 많이 되는 STEP 1이었습니다. 그렇다보니 이 방법이 좋은지 저 방법이 좋은지 고민하게 되고, 선택한 방법에는 이유가 모두 있다고 생각합니다. 이유가 타당하지 않을지도 모르니까 이상한 것 같은 부분은 최대한 알려주시면 좋겠습니다! 🙌

#### 함수별 기능
- `makeJuice`: 열거형 Juice의 case를 전달인자로 받아 주스를 만드는 메서드
- `add`: 클래스 FruitStore의 프로퍼티 inventory 내부 값에 전달인자를 더해주는 메서드
- `subtract`: 클래스 FruitStore의 프로퍼티 inventory 내부 값에서 전달인자를 빼주는 메서드
- `changeStock`: add, subtract, makeJuice 등에서 호출돼 클래스 FruitStore의 프로퍼티 inventory 내부 값을 변경시켜주는 메서드
- `checkStockOfFruits`: 레시피에 포함된 과일의 재고가 적절한지 확인하는 메서드

### 코드를 작성할 때 고민되었던 점
- 무엇보다 네이밍이 정말 고민이었습니다. 적절한 이름 짓는 게 진짜 너무 어려워요😅 최대한 상의하고 고민해가며 네이밍을 결정하였는데, 이상한 점은 말해주시면 감사하겠습니다.
- 클래스 FruitStore 내부의 프로퍼티 inventory를 딕셔너리로 결정하였기 때문에, 과일의 현재 수량에 접근하려고 하면 항상 옵셔널인 것이 문제였습니다. add, subtract 등의 메서드에서 이 값에 접근할 때 닐 병합 연산자를 사용할지 고민을 하였는데, 닐 병합연산자를 사용한다면 nil일 때 해당 과일의 잔여 수량을 0으로 설정해야해서 이 부분이 자연스럽지 않았습니다. 옵셔널 바인딩을 통해 인벤토리에 해당 과일 키가 있는지 구분하여 추가하는 방식으로 고민해본 결과 0 등의 매직 넘버를 사용하지 않고 해결할 수 있었습니다.
- 두 과일을 사용하는 딸기바나나, 망고키위 주스의 경우 한 과일이 재고가 충분하더라도 다른 과일이 불충분하면 만들 수 없어야 했습니다. 이 방법을 구현하기 위해 클래스 FruitStore 내부에 checkStockOfFruits 메서드를 생성해 실행한 결과 재료가 되는 과일 배열이 모두 충분할 때 쥬스를 만들게 했습니다.
- 클래스 FruitStore는 각 과일의 초기 재고가 10개로 요구되어 있으나, 꼭 모든 과일을 넣지 않을 수도 있고, 과일 별로 수량을 맞추지 않을 수도 있다고 생각해 원하는 값으로도 초기화 할 수 있도록 두 가지 경우로 이니셜라이저를 구분하였습니다.
- 쥬스메이커 인스턴스가 쥬스를 만들 때 canSubtract 메서드로 확인해서 오류를 미연에 방지하기 때문에 보통의 경우 오류가 없을 것이라고 생각했지만, 쥬스메이커를 거치지않고 인스턴스 fruitStore의 add, subtract에 직접 접근하는 경우가 있을 수 있다고 생각해 해당 메서드에 오류 처리를 하였습니다. 하지만 결국 makeJuice 메서드도 오류를 캐치해야한다고 생각해 지금의 checkSctockOfFruits 메서드를 만들고, makeJuice 메서드 내부에서 try 키워드가 두번 사용되지 않도록 과일 재고 수량을 변경하는 기능을 changStock 메서드로 분리해 add, subtract, makeJuice에서 호출하게 했습니다.
~~~swift
    // 변경 전
    func makeJuice(of juice: Juice) {
        for material in juice.recipe {
            guard fruitStore.canSubtract(amount: material.amount, of: material.fruit) else {
                return
            }
        }
        for material in juice.recipe {
            do {
                try fruitStore.subtract(amount: material.amount, of: material.fruit)
            } catch FruitStoreError.wrongAmount {
                print(FruitStoreError.wrongAmount.failureReason)
            } catch FruitStoreError.notInFruitList {
                print(FruitStoreError.notInFruitList.failureReason)
            } catch {
                print(error)
            }
        }
    }

    // 1차 변경
    func makeJuice(of juice: Juice) {
        do {
            try fruitStore.checkStockOfFruits(in: juice.recipe)
            for material in juice.recipe {
                try? fruitStore.subtract(amount: material.amount, of: material.fruit)
            }
        } catch FruitStoreError.wrongAmount {
            print(FruitStoreError.wrongAmount.failureReason)
        } catch FruitStoreError.notInFruitList {
            print(FruitStoreError.notInFruitList.failureReason)
        } catch FruitStoreError.outOfStock {
            print(FruitStoreError.outOfStock.failureReason)
        } catch {
            print(error)
        }
    }

    // 최종
    func makeJuice(of juice: Juice) {
        do {
            try fruitStore.checkStockOfFruits(in: juice.recipe)
            for material in juice.recipe {
                fruitStore.changeStock(of: material.fruit, amount: -material.amount)
            }
        } catch FruitStoreError.wrongAmount {
            print(FruitStoreError.wrongAmount.failureReason)
        } catch FruitStoreError.notInFruitList {
            print(FruitStoreError.notInFruitList.failureReason)
        } catch FruitStoreError.outOfStock {
            print(FruitStoreError.outOfStock.failureReason)
        } catch {
            print(error)
        }
    }
~~~

### 조언을 얻고 싶은 부분
- 쥬스를 만드는 데 필요한 과일의 수량만큼 재고를 감소하기 전에 레시피에 포함된 과일의 재고가 적절한지 checkStockOfFruits 메서드로 확인을 해주었습니다.
- checkStockOfFruit 메서드로 한 번 확인을 해주었기 때문에 subtract 메서드 대신 changeStock 메서드를 사용했습니다.
- subtract 메서드는 직접 호출할 때 적절한 인자인지 등을 체크해준 뒤 changeStock 메서드로 재고를 감소해주는 메서드입니다.
- changeStock 메서드는 별도의 확인 없이 재고를 조정해주는 메서드입니다.
- 이런 식으로 메서드를 분리해서 사용한 것에 대해 조언을 얻고 싶습니다. 더 좋은 방법이 있는지 궁금합니다.
~~~swift
    func subtract(amount: Int, of fruit: Fruit) throws {
        guard amount > 0 else {
            throw FruitStoreError.wrongAmount
        }
        guard let currentStock = inventory[fruit] else {
            throw FruitStoreError.notInFruitList
        }
        guard currentStock >= amount else {
            throw FruitStoreError.outOfStock
        }
        changeStock(of: fruit, amount: -amount)
    }
    
    func changeStock(of fruit: Fruit, amount: Int) {
        if let currentStock = inventory[fruit] {
            inventory[fruit] = currentStock + amount
        }
    }
~~~

- 54bdc31 커밋까지 진행했을 시, 오류 처리를 구현하지 않아도 모든 기능이 정상 작동을 하는데, 과연 재고 부족 등이 에러인건지, 꼭 오류 처리를 해줘야하는지 조언을 구하고 싶습니다. 예상되는 조건을 모두 처리했는데 그것이 에러라고 할 수 있는 것인지 궁금합니다.

Gundy93 committed:
feat: 과일을 나타내는 열거형 Fruit 추가

junho15 committed:
feat: FruitStore의 inventory 프로퍼티와 init 추가

Gundy93 committed:
feat: 재고를 추가하는 add 메서드 추가

junho15 committed:
feat: 재고를 감소시키는 subtract 메서드 추가

Gundy93 committed:
feat: 쥬스를 나타내는 열거형 Juice 추가

junho15 committed:
feat: 재고가 충분한 지 확인하는 canSubtract 메서드 추가

Gundy93 committed:
feat: 쥬스를 만드는 makeJuice 메서드 추가

junho15 committed:
refactor: 열거형 Juice, 클래스 FruitSotre, 구조체 JuiceMaker 수정

- 열거형 Juice의 Materail 구조체 및 recipe 프로퍼티 수정
- 클래스 FruitStore의 init 추가, self 문구 제거
- 구조체 JuiceMaker의 fruitStore 프로퍼티 기본값 제거, makeJuice 메서드 수정

Gundy93 committed:
refactor: 구조체 JuiceMaker의 fruitStore 프로퍼티 접근제어, 클래스 FruitStore의 inventory 프로퍼티 설정자 접근제어

junho15 committed:
feat: 에러 프로토콜을 준수하는 열거형 FruitStoreError 추가

Gundy93 committed:
refactor: 오류처리(throw, try-catch)구문 추가, 구조체 JuiceMaker init 추가

junho15 committed:
refactor: FruitStoreError, FruitStore, JuiceMaker 수정

- FruitStoreError에 outOfStock 케이스 추가
- FruitStore에 canSubtract 메서드를 삭제 checkStockOfFruits 메서드 추가
- FruitStore에 changeStock 메서드를 추가해서 add, subtract 메서드에서 사용
- JuiceMaker의 makeJuice 메서드 수정(canSubtract메서드 대신 checkStockOfFruits메서드로 확인, changeStock메서드로 재고 수정)

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
-
-// 과일 저장소 타입
 class FruitStore {
+    private(set) var inventory: Dictionary<Fruit, Int> = [:]
```
오..private set 을 사용하신 이유 여쭤봐도 될까요?!
틀렸다는건 아닙니다~

Gundy93 replied:
직접적으로 과일 수량을 변경하는 것을 차단하지만, 재고가 몇개인지 확인할 수 있어야 추가할 수량을 결정할 수 있다고 생각해 설정자만 접근 수준을 private로 결정했습니다.

junbangg replied:
`fruitStore.inventory[fruit]` 이런식으로 확인한다는 것이죠??
그것도 충분히 가능하고~ `private(set)`을 이용해서 잘 구현도 해주셨네요👍👍👍👍 

나중에 객체지향프로그래밍에 대한 공부를 더 하시겠지만 차라리 `inventory` 를 아예 숨겨버리고
`fruitStore.checkStock(of: fruit)` 를 하는 것이 더 좋아보여요 저는!
개인적인 생각쓰~~

Gundy93 replied:
만약 그렇게 한다면 `지금 add , subtract, changeStockOfFruits 메서드들을 보면 셋다 재고를 체크하는 로직을 갖고 있어요`라는 코멘트도 동시에 해결되는군요! 준호와 잘 상의해보겠습니다!

Gundy93 replied:
01d0f3a 커밋에서 해당 사항 수정하였습니다.
inventory 프로퍼티를 private으로 변경하고, 현재 재고를 가져오는 getCurrentStock 메서드를 추가했습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
 class FruitStore {
+    private(set) var inventory: Dictionary<Fruit, Int> = [:]
+    
+    init(inventory: Dictionary<Fruit, Int>) {
```
생성자로 `inventory` 를 잘 주입 해주셨네요..!
이렇게 주입하고있는데 위에서는 초기화를 하고 있는 이유가 따로 있을까요?????(8번째 줄)

Gundy93 replied:
작업 과정에서 실수로 초기값이 들어간 것 같습니다. 여러번 확인했는데 이상하다고 깨닫지 못했던 것 같습니다🥲

Gundy93 replied:
2489cc6 커밋에서 해당 사항 수정하였습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+    
+    init(initialStock: Int) {
+        for fruit in Fruit.allCases {
+            inventory[fruit] = initialStock
```
과일마다 만약에 초기 재고가 달라지는 상황이 생기면 어떻게 할지도 한번 생각해보면 좋을것 같은데요?!
Ex: 딸기:3개, 바나나: 15개

Gundy93 replied:
10번 라인의 `init(inventory: Dictionary<Fruit, Int>)` 이니셜라이저가 해당 역할을 한다고 생각합니다! 저희도 과일 개수를 저마다 다르게 생성하고 싶을 경우가 있을거라고 생각해 만든 이니셜라이저였습니다.

junbangg replied:
inventory를 주입할 수 있는 옵션을 제공하는 것에 대해서 정확히 말씀 드리려고 한건데
바로 위에서 이미 해주셨네요ㅋㅋ정신이 없었네요 굳굳👍👍

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+        }
+    }
+    
+    func add(amount: Int, of fruit: Fruit) throws {
```
파라미터 레이블로 가독성을 높여주셨군요..!
가독성에 대한 고민을 많이 하신것 같네요👍👍

조금 더 구체적인 네이밍도 가능할것 같아요..!
`increaseStock(of fruit: Fruit, by amount: Int)`
아니면 지금 네이밍 조금 변경해서
`increase(_ fruit: Fruit, by amount: Int)`
요런 느낌도 내볼 수 있을것 같고..
근데 이건 어디까지나 제 의견/스타일이고 꼭 따를 필요는 없습니다~ 그냥 요런 네이밍도 가능할것 같다~ 정도 ?

지금도 가독성 좋다고 생각해요..! 👍👍

Gundy93 replied:
네이밍은 항상 어려운 것 같습니다...말씀해주신 네이밍도 정말 좋은 것 같은데, 준호와 상의해서 결정하겠습니다! 😃

Gundy93 replied:
2489cc6 커밋에서 해당 네이밍 수정하였습니다. 추가로 `subtract`도 `increase`에 맞춰 `decrease`로 변경하였습니다! 😃

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+    
+    func add(amount: Int, of fruit: Fruit) throws {
+        guard amount > 0 else {
+            throw FruitStoreError.wrongAmount
```
흠 `wrongAmount` 라는 네이밍이 조금 모호한것 같네요 개인적으로.
`wrongAmount` 를 봤을때 저는 `틀린/잘못된 amount` 로 읽히는데
`틀린 amount` 보다는 `invalid` 같은 단어를 쓰는거에 대해서는 어떻게 생각하세요??

두 네이밍의 차이점이 혹시 느껴지시나요?

Gundy93 replied:
숫자가 양수가 아닐 경우 `잘못된 amount`라고 생각해서 해당 네이밍을 결정했었는데, 이제와서 생각해보니 타입을 Int로 설정해놓고 0 및 음의 정수값을 틀렸다고 표현하는 게 조금 이상하긴 한 것 같네요...`유효하지 않은/효력이 없는 amount`가 더 적절한 것 같습니다.

Gundy93 replied:
2489cc6 커밋에서 해당 사항 `invalidAmount`로 수정하였습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+        guard inventory[fruit] != nil else {
+            throw FruitStoreError.notInFruitList
+        }
+        changeStock(of: fruit, amount: amount)
```
`guard` 문으로 딱딱 나눠서 오류 처리한거 칭찬해요..! :)

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+        changeStock(of: fruit, amount: -amount)
+    }
+    
+    func changeStock(of fruit: Fruit, amount: Int) {
```
`func changeStock(of fruit: Fruit, by amount: Int)` 로 하면
위에서 사용할때
`changeStock(of: fruit, by: -amount)` 
처럼 사용할 수 도 있을것 같네요?!

Gundy93 replied:
아...! 아규먼트 레이블을 조금 더 신경쓰겠습니다! 바뀐 표현이 호출시에 더 깔끔해 보이네요! of 까지는 생각했는데 amount의 경우를 생각하지 못한 것 같습니다.

Gundy93 replied:
2489cc6 커밋에서 해당사항 `changeStock(of fruit: Fruit, by amount: Int)`로 수정하였습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+    func changeStock(of fruit: Fruit, amount: Int) {
+        if let currentStock = inventory[fruit] {
+            inventory[fruit] = currentStock + amount
+        }
```
가독성 좋은것 같아요! :)

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+            guard currentStock >= material.amount else {
+                throw FruitStoreError.outOfStock
+            }
+        }
```
레시피의 재료에 대한 재고 체크를 하고 있는거군요???
깔끔한것 같아요!
근데 `Material` 을 쓰신 이유는 `재료` 를 나타내기 위한건가요??
그렇다면 `Material` 보다는 `Ingredient` 어떠세요??

junbangg replied:
레시피의 모든 재료의 재고를 체크하고 있는 메서드로 보이는군요
근데! 만약에 그냥 딸기라는 재료 하나의 재고를 체크 하고싶을때는 어떻게 해야될까요,,?🤔

junbangg commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,40 @@
+        let amount: Int
+    }
+    
+    var recipe: [Material] {
```
`recipe` 가 재료의 배열이다?
이런거죠?
아아아주 좋습니다:) 
`Material` 보다는 `Ingredient` 가 직관적일것 같네요 개인적으로

Gundy93 replied:
찾아보니 재료를 나타내는 단어 'material', 'ingredient', 'nutrient' 중 식재료를 나타내는 것은 'ingredient'였군요! 저도 `Ingredient`로 바꾸는 것이 더 합당한 것 같습니다. 😄

Gundy93 replied:
2489cc6 커밋에서 해당 사항 수정하였습니다. `material` 혹은 `Material`이 사용된 부분은 전부 `ingredient`, `Ingredient`로 수정하였습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,9 +4,27 @@
+        self.fruitStore = fruitStore
+    }
     
+    func makeJuice(of juice: Juice) {
```
흠 메서드를 읽어보면
`make juice of juice` 네요??
조금 더 개선해볼 수 있을것 같아요..!

Gundy93 replied:
호출 시에는 `makeJuice(of: .strawberry)` 등의 형태가 될 것이라고 생각해 괜찮다고 여겼는데 정의부에서는 그렇지 않아보이네요. 처음에는 `makeJuice(of flavor: Juice)`로 상의한 적도 있습니다. 조금 더 고민과 상의를 거쳐서 수정해보겠습니다!

Gundy93 replied:
2489cc6 커밋에서 해당 사항 수정하였습니다. `make(_ juice: Juice)로 수정하여 메서드를 읽을 땐 `make juice`, 호출할 땐 `make strawberry juice` 등으로 읽히게 하였습니다! 😃

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStoreError.swift:
```diff
@@ -0,0 +1,23 @@
+    case notInFruitList
+    case outOfStock
+    
+    var failureReason: String {
```
에러 메세지를 나타내는 연산속성을 만들어주신거 좋습니다!
`description` , `localizedDescription`  으로도 많이 사용하시는것 같아요 :) 
근데 이건 살짝 컨벤션 이야기이긴 해서 지금은 너무 신경 안쓰셔도 될것 같아요! 어차피 나중에 보게 되실것 같아요~

`failureReason` 도 신경을 쓴 네이밍 같네요

Gundy93 replied:
처음에는 `localizedDescription`으로 하려다 저희의 에러 메시지는 '서술'보다는 짤막한 '이유'를 알린다고 생각해 `failureReason`로 수정하였었습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+        for material in recipe {
+            guard let currentStock = inventory[material.fruit] else {
+                throw FruitStoreError.notInFruitList
+            }
```
레시피의 재료에 대한 재고 체크를 하고 있는거군요???
근데 `Material` 을 쓰신 이유는 `재료` 를 나타내기 위한건가요??
그렇다면 `Material` 보다는 `Ingredient` 어떠세요??

junbangg replied:
레시피의 모든 재료의 재고를 체크하고 있는 메서드로 보이는군요
근데! 만약에 그냥 딸기라는 재료 하나의 재고를 체크 하고싶을때는 어떻게 해야될까요,,?🤔

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,57 @@
+        }
+    }
+    
+    func checkStockOfFruits(in recipe: [Juice.Material]) throws {
```
레시피의 재료에 대한 재고 체크를 하고 있는거군요???
근데 Material 을 쓰신 이유는 재료 를 나타내기 위한건가요??
그렇다면 Material 보다는 Ingredient 어떠세요??

junbangg replied:
레시피의 모든 재료의 재고를 체크하고 있는 메서드로 보이는군요
근데! 만약에 그냥 딸기라는 재료 하나의 재고를 체크 하고싶을때는 어떻게 해야될까요,,?🤔

Gundy93 replied:
`그렇다면 Material 보다는 Ingredient 어떠세요??`의 제 의견은 상부 코멘트에 달았습니다!
현재는 1개 또는 2개의 재료만을 사용하는 주스 레시피지만, 만약 딸키바주스(딸기+키위+바나나)같은 레시피가 추가되더라도 배열이라면 재료의 가짓수와 무관하게 사용할 수 있다고 생각했습니다. 배열 내부에 요소가 1개여도 적절한 작동을 하고 반복문이 종료될테니까요! 주스를 만들지 않고 재고 체크만을 하는 경우라면 저장 프로퍼티 `inventory`에서 해당 과일의 재고를 체크하면 어떤가 생각해 `inventory`를 `private(set)`으로 설정했습니다.

Gundy93 replied:
01d0f3a 커밋에서 해당 사항 수정하였습니다.
재료 하나의 재고가 충분한 지 확인하는 checkInventoryHasStock 메서드를 추가했습니다.
checkStockOfFruits 메서드 이름을 checkStockOfIngredients로 변경하고, 메서드 내부의 재고 확인 기능을 checkInventoryHasStock 메서드로 처리했습니다.

junbangg commented:
수고 많으셨습니다~~
가독성에 대한 고민의 흔적들이 많이 느껴진 코드였어요!
그리고 자세하게 `PR` 적어주신거 아주 칭찬합니다!
고민을 한 과정들을 정리해주신게 인상적이었어요
아주 좋은 습관이라고 생각합니다. 앞으로도 꾸준히 고민한 과정들을 기록하는 습관을 갖는다면 정말 큰 자산이 될거라고 생각해요👍

## 그리고 질문해주신 내용들을 살펴보자면~
### 1. 가독성
가독성/네이밍 아주 중요하고 아주 어렵죠~~ 고민 많이 하신것 같아요! 그게 중요한거죠:)
이부분은 댓글 참고해주시면 될것 같아요~~

### 2. 메서드 분리
역할별로 분리하려고 노력하신 점 칭찬합니다 :) 
그리고 재고를 추가하고 빼는 로직을 하나의 메서드로 재사용하려고 한 노력도 아주 좋았어요~!

근데 조금은 개선할 수 있을것 같아요!

지금 `add` , `subtract`, `changeStockOfFruits` 메서드들을 보면 셋다 재고를 체크하는 로직을 갖고 있어요

재고를 체크하는 메서드를 아예 만들어버리는건 어떠신가요??

작은 조언을 하나 더 드리자면 지금 `FruitStore` 라는 친구가 `inventory` 라는 프로퍼티를 가지고 정말 다양한 일들을 하고있죠?
재고를 체크하고, 추가하고, 빼고 말이죠

그렇다면 아예 그냥 `Inventory` 라는 타입을 추가로 만들어서 `FruitStore` 이 갖고 있게 하는건 어떨까요?? 
이건 조언일 뿐이고 꼭 그렇게 할 필요는 없습니다~ 두분이서 한번 같이 생각해보시라~~이런거죠 :)

### 3. "과연 재고 부족 등이 에러인건지, 꼭 오류 처리를 해줘야하는지 조언을 구하고 싶습니다."
두분께서는 오류 처리가 필요한 이유가 어떤것인지 궁금합니다:) 
또한~ 여기서 오류처리를 빼버리면 어떤 상황이 발생할 수 있는지 한번 이야기해보면 좋을것 같아요 ! (오류를 리턴하지 않으면 뭐가 대신 리턴 될까요?)

수고하셨습니다:)

Gundy93 committed:
refactor: 메서드와 열거형 케이스, 타입 네이밍 수정

- 메서드
	- add -> increaseStock(of fruit: Fruit, by amount: Int)
	- subtract -> decreaseStock(of fruit: Fruit, by amount: Int)
	- changeStock(of fruit: Fruit, amount: Int) -> changeStock(of fruit: Fruit, by amount: Int)
	- makeJuice -> make(_ juice: Juice)
- 열거형 케이스
	- wrongAmount -> invalidAmount
	- 모든 주스 -> 뒤에 Juice 덧붙임
- 타입
	- Material -> Ingredient

junho15 committed:
refactor: getCurrentStock, checkInventoryHasStock 메서드 추가 및 checkStockOfFruits 메서드 이름 변경, 기능 분리

- inventory 프로퍼티를 private(set) 에서 private로 변경
- 현재 재고를 가져오는 getCurrentStock 메서드 추가
- 재고가 충분한 지 확인하는 checkInventoryHasStock 메서드 추가
- checkStockOfFruits 메서드 이름을 checkStockOfIngredients로 변경
- checkStockOfIngredients 메서드 내부의 재고 확인 기능을 checkInventoryHasStock 메서드로 분리

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,59 @@
     
+    func getCurrentStock(of fruit: Fruit) throws -> Int {
+        guard let currentStock = inventory[fruit] else {
+            throw FruitStoreError.notInFruitList
```
`inventory[fruit]` 접근했을때 `nil` 이 나올 경우에 오류가 던져지게 잘 해주셨는데
에러 이름이 왜 `.notInFruitList` 인가요?? 
위에 `recipe` 를 나타내는 과일 배열이랑 혼돈이 생길 수 있을것 같아요!

Gundy93 replied:
인벤토리에서 보유하고있는 과일목록에 없는 과일을 체크하려고 할 경우 없다! 라고 한 것이었습니다. `FruitStore` 타입의 인스턴스를 초기화할 때, 모든 과일 케이스를 초기 재고로 갖지 않고 일부 과일만을 가지는 경우가 발생할 수 있다고 생각해 그렇게 했었습니다.

junho15 replied:
04e27db 커밋에서 해당 사항 수정했습니다.
notInFruitList를 notInInventoryFruitList로 바꿔서 네이밍을 구체화 했습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,59 @@
+        guard amount > 0 else {
+            throw FruitStoreError.invalidAmount
+        }
+        try _ = getCurrentStock(of: fruit)
```
요 부분은 조금 개선할 필요가 있어보여요!
이름을 봤을때는 `getCurrentStock(of: fruit)` 가 과일의 재고값을 가져오는 역할의 메서드인거 같은데 
지금 실제로 사용되고있는 방법은 이름과 조금 다른거 같아요!
`getCurrentStock` 가 오류를 던지고/안던지고에 따라서 재고가 있고 없고를 판별하고 있는거 같은데 
혹시 이렇게 구성한 이유가 있을까요??  `Bool` 타입을 반환하는 메서드를 사용하지 않는 이유가 따로 있는걸까요??

p.s 
지금 스텝에서 중요한 포인트는 아니지만
와일드 카드(`_`) 를 사용하는 방법도 있지만 
조금 더 깔끔하게 반환값을 안받는 방법도 있을것 같아요
`@discardableResult` 한 번 검색해주세용

Gundy93 replied:
재고가 있고 없고를 판별함과 동시에 Int를 반환하여, 49번 줄의 `getCurrentStock` 메서드를 호출하는 `checkInventoryHasStock` 메서드에서 옵셔널 바인딩 코드가 중복되지 않고 바로 값 비교를 하게끔 하려는 의도였습니다. 하지만 알라딘의 말을 듣고 보니 '재고의 체크'와 '재고값의 반환'이라는 두 기능이 담긴 것 같습니다...@discardableResult에 대해서도 같이 공부해보겠습니다.

junho15 replied:
3f7151e 커밋에서 해당 사항 수정했습니다.
재고값을 확인할 필요가 없는 부분에서는 창고 내부 과일 목록에 있는지 확인해 Bool 타입을 반환하는 isInventoryFruitsListHas 메서드 추가해서 활용했습니다.

재고를 가져오는 메서드와 재고 목록에 있는지 확인하는 메서드로 분리해서 @discardableResult는 이번엔 사용하지 않았습니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -4,9 +4,59 @@
-
-// 과일 저장소 타입
 class FruitStore {
+    private var inventory: Dictionary<Fruit, Int>
```
`private` 로 은닉화를 잘 해주셨는데 밑에 마찬가지로 `private` 시킬 수 있는 메서드는 없으려나요?!

Gundy93 replied:
아! 메서드를 분리 및 추가하면서 은닉화를 하지 않았던 것 같습니다. 이 부분도 전반적으로 다시 확인해보겠습니다! 💪

junho15 replied:
04e27db 커밋에서 해당 사항 수정했습니다.
(기존)checkInventoryHasStock, (신규)isInventoryFruitsListHas 메서드 은닉화했습니다.

Gundy93 committed:
feat: 창고 내부 과일 목록에 있는지 확인해 Bool값을 반환하는 isInventoryFruitsListHas 메서드 추가

- increaseStock 메서드 내부에서 활용

junho15 committed:
refactor: notInFruitList를 notInInventoryFruitList로 네이밍 구체화, checkInventoryHasStock, isInventoryFruitsListHas 메서드 은닉화

Gundy93 committed:
docs: README.md 생성
